lib/adapters/show-document-adapter.ts

```diff
@@ -0,0 +1,43 @@
+import * as url from 'url';
+import { shell } from 'electron';
+import {
+  LanguageClientConnection,
+  ShowDocumentParams,
+  ShowDocumentResult,
+} from '../languageclient';
+import Convert from '../convert';
+
+/** Public: Adapts the window/showDocument command to Atom's text editors or external programs. */
+export default class ShowDocumentAdapter {
+  /**
+   * Public: Attach to a {LanguageClientConnection} to recieve requests to show documents.
+   */
+  public static attach(connection: LanguageClientConnection): void {
+    connection.onShowDocument(ShowDocumentAdapter.onShowDocument);
+  }
+
+  /**
+   * Public: Show a notification message with buttons using the Atom notifications API.
+   *
+   * @param params The {ShowDocumentParams} received from the language server
+   *   indicating the document to be displayed as well as other metadata.
+   */
+  public static async onShowDocument(params: ShowDocumentParams): Promise<ShowDocumentResult> {
+    if (!params.external) {
+      const editor = await atom.workspace.open(params.uri, {
+        activateItem: params.takeFocus,
+        activatePane: params.takeFocus,
+        pending: true,
+        initialLine: params.selection?.start.line ?? 0,
+        initialColumn: params.selection?.start.character ?? 0,
+      });
+      if (params.selection != null) {
+        editor.selectToBufferPosition(Convert.positionToPoint(params.selection.end));
+      }
+      return { success: true };
+    }
+
+    shell.openExternal(url, { activate: params.takeFocus })
```

Comment:
Maybe instead of opening it here we could send it to the client and the client will handle opening the url?

Reply:
How can we send it? Define a callback on `AutoLanguageClient`? Moreover, that'd mean doing the same thing over and over in every language client, we should avoid this if possible

Reply:
I am fine with opening it here.

Reply:
I'm not sure if I am reading the spec correctly but it seems according to the spec that the uri is supposed to be a document that should be opened in Atom not externally.

Reply:
```suggestion
    shell.openExternal(params.uri, { activate: params.takeFocus })
```

Reply:
Yep, didn't notice it in because of the TypeScript error. Fixed now.